planner/core/task.go

```diff
@@ -126,46 +138,139 @@ func (p *PhysicalApply) attach2Task(tasks ...task) task {
+	innerCPUCost += outerCnt * distinctFactor * innerCnt * cpuFactor
+	innerConcurrency := float64(p.ctx.GetSessionVars().IndexLookupJoinConcurrency)
+	cpuCost += innerCPUCost / innerConcurrency
+	// Cost of probing hash table in main thread.
+	numPairs := outerCnt * innerCnt
+	if p.JoinType == SemiJoin || p.JoinType == AntiSemiJoin ||
+		p.JoinType == LeftOuterSemiJoin || p.JoinType == AntiLeftOuterSemiJoin {
+		if len(p.OtherConditions) > 0 {
+			numPairs *= 0.5
+		} else {
+			numPairs = 0
+		}
 	}
-	cst := lCnt * netWorkFactor
-	batchSize := p.ctx.GetSessionVars().IndexJoinBatchSize
-	cst += lCnt * math.Log2(math.Min(float64(batchSize), lCnt)) * 2
-	cst += lCnt / float64(batchSize) * netWorkStartFactor
-	return cst
+	probeCost := numPairs * cpuFactor
+	// Cost of additional concurrent goroutines.
+	cpuCost += probeCost + (innerConcurrency+1.0)*concurrencyFactor
+	// Memory cost of hash tables for inner rows. The computed result is the upper bound,
+	// since the executor is pipelined and not all workers are always in full load.
+	memoryCost := innerConcurrency * (batchSize * distinctFactor) * innerCnt * memoryFactor
+	// Cost of inner child plan, i.e, mainly I/O and network cost.
+	innerPlanCost := outerCnt * innerTask.cost()
+	return outerTask.cost() + innerPlanCost + cpuCost + memoryCost
 }
 
-func (p *PhysicalHashJoin) getCost(lCnt, rCnt float64) float64 {
-	smallTableCnt := lCnt
+// GetCost computes cost of hash join operator itself.
+func (p *PhysicalHashJoin) GetCost(lCnt, rCnt float64) float64 {
+	innerCnt, outerCnt := lCnt, rCnt
 	if p.InnerChildIdx == 1 {
-		smallTableCnt = rCnt
+		innerCnt, outerCnt = rCnt, lCnt
+	}
+	// Cost of building hash table.
+	cpuCost := innerCnt * cpuFactor
+	memoryCost := innerCnt * memoryFactor
+	// Number of matched row pairs regarding the equal join conditions.
+	helper := &fullJoinRowCountHelper{
+		cartesian:     false,
+		leftProfile:   p.children[0].statsInfo(),
+		rightProfile:  p.children[1].statsInfo(),
+		leftJoinKeys:  p.LeftJoinKeys,
+		rightJoinKeys: p.RightJoinKeys,
+		leftSchema:    p.children[0].Schema(),
+		rightSchema:   p.children[1].Schema(),
+	}
+	numPairs := helper.estimate()
+	// For semi-join class, if `OtherConditions` is empty, we already know
+	// the join results after querying hash table, otherwise, we have to
+	// evaluate those resulted row pairs after quering hash table; if we
+	// find one pair satisfying the `OtherConditions`, we then know the
+	// join result for this given outer row, otherwise we have to iterate
+	// to the end of those pairs; since we have no idea about when we can
+	// terminate the iteration, we assume that we need to iterate half of
+	// those pairs in average.
+	if p.JoinType == SemiJoin || p.JoinType == AntiSemiJoin ||
+		p.JoinType == LeftOuterSemiJoin || p.JoinType == AntiLeftOuterSemiJoin {
+		if len(p.OtherConditions) > 0 {
+			numPairs *= 0.5
+		} else {
+			numPairs = 0
```

Comment:
Why set to 0? When it's semi join, we still need to probe it?

Reply:
If it is semi join and `OtherConditions` is empty, we would only probe the hash table, which is very cheap, and there is no filter evaluation and row concatenation. Please see comments above and below for explanation. Here `probeCost` does not refer to the cost of probing hash table, it means the cost of evaluating the row pairs after probing hash table.